scripts/http-security-headers.nse

```diff
@@ -0,0 +1,305 @@
+--
+--
+-- @xmloutput
+-- <elem>HSTS is configured</elem>
+-- <elem key="Header">Strict-Transport-Security: max-age=31536000</elem>
+-- <elem>HPKP is configured</elem>
+-- <elem key="Header">Public-Key-Pins: pin-sha256="d6qzRu9zOECb90Uez27xWltNsj0e1Md7GkYYkVoZWmM="; report-uri="http://example.com/pkp-report"; max-age=10000; includeSubDomains</elem>
+-- <elem>X-Frame-Options is configured</elem>
+-- <elem key="Header">X-Frame-Options: DENY</elem>
+-- <elem key="Description">The browser must not display this content in any frame.</elem>
+-- <elem>X-XSS-Protection is configured</elem>
+-- <elem key="Header">X-XSS-Protection: 1; mode=block</elem>
+-- <elem key=Description>Rather than sanitize the page, when a XSS attack is detected, the browser will prevent rendering of the page.</elem>
+-- <elem>X-Content-Type-Options is configured.</elem>
+-- <elem key="Header">X-Content-Type-Options: nosniff</elem>
+-- <elem key="Description">Will prevent the browser from MIME-sniffing a response away from the declared content-type.</elem>
+-- <elem>Content-Security-Policy is configured.</elem>
+-- <elem key="Header">Content-Security-Policy: script-src 'self'</elem>
+-- <elem key="Description">Loading policy for all resources type in case of a resource type dedicated directive is not defined (fallback).</elem>
+-- <elem>X-Permitted-Cross-Domain-Policies are configured.</elem>
+-- <elem key="Header">X-Permitted-Cross-Domain-Policies: none</elem>
+-- <elem key="Description">No policy files are allowed anywhere on the target server, including this master policy file.</elem>
+--
+-- @args http-security-headers.path The URL path to request. The default path is "/".
+
+author = {"Icaro Torres", "Vinamra Bhatia"}
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "safe"}
+
+portrule = shortport.port_or_service({80,443}, "http", "tcp")
+
+local function fail (err) return stdnse.format_output(false, err) end
+
+action = function(host, port)
+  local path = stdnse.get_script_args(SCRIPT_NAME..".path") or "/"
+  local response
+  local output_info = {}
+  local hsts_header = {}
+  local hpkp_header = {}
+  local xframe_header = {}
+  local x_xss_header = {}
+  local x_content_type_header = {}
+  local csp_header = {}
+  local x_cross_domain_header = {}
+  local cookie = {}
+  local req_opt = {redirect_ok=function(host,port)
+    local c = 2
+    return function(uri)
+      if ( c==0 ) then return false end
+      c = c - 1
+      return true
+    end
+  end}
+
+  response = http.head(host, port, path, req_opt)
+
+  output_info = stdnse.output_table()
+
+  if response == nil then
+    return fail("Request failed")
+  end
+
+  if response.rawheader == nil then
+    return fail("Response didn't include a proper header")
+  end
+
+  for _, line in pairs(response.rawheader) do
+    if line:match("[Ss]trict.[Tt]ransport.[Ss]ecurity") then
```

Comment:
For header matches in this loop, it would be safer to bind the regexes to the start of the line. That prevents unexpected header values from being misinterpreted as header names.